mutation_reader.cc

```diff
@@ -1048,6 +1050,20 @@ void shard_reader::remote_reader::adjust_partition_slice() {
     auto ranges = _slice_override->default_row_ranges();
     query::trim_clustering_row_ranges_to(*_schema, ranges, last_ckey);
 
+    if (_last_position_in_partition_is_prefix) {
```

Comment:
If we stopped at a clustering row, we don't need to filter, because slicing will make sure we won't see any old fragments. In that case we could set `drop_needed` to false.

But I wonder if we ever need to filter. We could do this instead:

1) change `trim_clustering_row_ranges_to()` to work with `position_in_partition`, and make the resulting range include the position rather than exclude it like currently. We can always include a position_in_partition, but we cannot always exclude it (we cannot exclude the start of a prefix, which is the problem we have here). We can avoid the ugly `find_if` this way, and also this will allow us to avoid filtering by also doing this change:

We should set `_last_position_in_partition` (which now would be included rather than excluded) not from `buffer.back()` in `update_last_position()`, but in `ensure_buffer_contains_all_fragments_for_last_pos()` in the following way:
  - if we stopped at a clustering row with key `k`, set `_last_position_in_partition` to `after(key)`.
  - if we had to consume tombstones and stopped by peeking a fragment with key `k2`, set `_last_position_in_partition` to `k2`.

Reply:
Thinking about it, we should also always trim range tombstones to _last_position_in_partition, since after recreation we may get range tombstones starting before the start of the slice, right?

Reply:
> If we stopped at a clustering row, we don't need to filter, because slicing will make sure we won't see any old fragments. In that case we could set drop_needed to false.

We can't because the `partition_start` and the `static_row` will be re-emitted again anyway.

> Thinking about it, we should also always trim range tombstones to _last_position_in_partition, since after recreation we may get range tombstones starting before the start of the slice, right?

Possibly, but with the current code those will not be emitted again (`should_drop_fragment()` will flag them up).

I like the `position_in_partition` idea. I'm already in the process of refactoring `trim_clustering_ranges_to()` because of the other issue I found after submitting this PR. The issue is, just FYI, that `trim_clustering_ranges_to()` is not prepared for a key that is a prefix, and in this condition it will eliminate ranges, whoose end bound key is a prefix of. For example, for the inputs:
```
range=[{0, 0}, {0, 99999})
key={0}
```
The range will be dropped.

In general prefixes are confusing because they do not represent a point in the key-space, rather a range in themselves, breaking code that is not prepared for this. The world would be a better place if we had prefixes replaced with full compounds with unset components pre-filled with min/max values for the type, accordingly.

Reply:
For `partition_start` and the `static_row` we need the old filtering, yes.

About dropping the tombstones, I think it is not correct to completely drop them, because after recreation their `end` may have been extended due to merging with later fragments.

yes, a prefix inside `clustering_range` has a different meaning than a prefix as a standalone key. The latter is positioned as if it was null-extended. While a prefix in the range means that the range either excludes or includes the whole prefix.

`position_in_partition` was supposed to solve this problem. You can always treat a range as a pair of p_i_p, and assume the range includes all p_i_p such that start < p <= end.

Reply:
I don't understand how the `end` could be extended after recreating the reader. We create the reader with monotonically narrowing `range` (moving only `start`, never `end`).

Reply:
A range deletion [1, 5] can be also represented as [1, 3) and [3, 5]. Before recreation you could have get [1, 3) then pause, and have [3, 5] yet to consume. But after recreation you may get [1, 5]. If you ignore it, the upper consumer will only know about [1, 3) and won't know that [3, 5] is also deleted.

Reply:
Right, unfortunately many (most) of our tests expect the exact representation of writes back that they provided and will fail for such a re-emitted range-tombstone.

Also, I found another problem. Consider reading a range bounded by two non-full prefixes (`-` denotes missing component):
```
[{0, -}, {2, -}]
```
Eventually we will arrive at a singular-looking range:
```
[{2, -}, {2, -}] == [{2, -}]
```
Now if the reader is recreated after having read some clustering row, the range remaining to read will look like this:
```
({2, 1}, {2, -}]
```
Depending on how you look this is an invalid or wrapping range, as `end` includes positions that are before `start`, although in practice it might just work without problems.

Reply:
It may look confusing, yes, becasue the meaning of the prefix is overloaded when put in a range. So you cannot just take clustering_row's key and put it in a range bound and hope it will mean that the range will start right after the key. Prefix keys have to be null-extended when put into a range.

Currently, the correct interpretation of a prefix inside a range is using prefix equality comparison. Every key which is prefixed by a prefix in the range's bound is considered equal to the bound. Then exclusiveness is used to decide whether equal things are in or out.

With position_in_partition there is arguably less confusion, since bound equality is the same as p_i_p equality.

Reply:
> Thinking about it, we should also always trim range tombstones to _last_position_in_partition, since after recreation we may get range tombstones starting before the start of the slice, right?

This should already be done by upstream readers. It is illegal for readers to emit fragments outside of the query range and slice. A violation of this was fixed not long ago, see https://github.com/scylladb/scylla/issues/4104.

Reply:
> We should set `_last_position_in_partition` (which now would be included rather than excluded) not from `buffer.back()` in `update_last_position()`, but in `ensure_buffer_contains_all_fragments_for_last_pos()` in the following way:
> 
>     * if we stopped at a clustering row with key `k`, set `_last_position_in_partition` to `after(key)`.
> 
>     * if we had to consume tombstones and stopped by peeking a fragment with key `k2`, set `_last_position_in_partition` to `k2`.

`ensure_buffer_contains_all_fragments_for_last_pos()` is not a reliable place to do this at, it is only invoked when needed (we return early when we are good), but we always need a `_last_position_in_partition`. I found another solution. v2 is coming soon.

Reply:
I can see that `ensure_buffer_contains_all_fragments_for_last_pos()` is always invoked by `fill_buffer()`.